cranelift/codegen/src/alias_analysis.rs

```diff
@@ -0,0 +1,387 @@
+        let opcode = func.dfg[inst].opcode();
+        if has_memory_fence_semantics(opcode) {
+            self.heap = inst.into();
+            self.table = inst.into();
+            self.vmctx = inst.into();
+            self.other = inst.into();
+        } else if opcode.can_store() {
+            if let Some(memflags) = func.dfg[inst].memflags() {
+                if memflags.heap() {
+                    self.heap = inst.into();
+                } else if memflags.table() {
+                    self.table = inst.into();
+                } else if memflags.vmctx() {
+                    self.vmctx = inst.into();
+                } else {
+                    self.other = inst.into();
+                }
+            } else {
+                self.heap = inst.into();
+                self.table = inst.into();
+                self.vmctx = inst.into();
+                self.other = inst.into();
+            }
+        }
+    }
+
+    fn get_last_store(&self, func: &Function, inst: Inst) -> PackedOption<Inst> {
+        if let Some(memflags) = func.dfg[inst].memflags() {
+            if memflags.heap() {
+                self.heap
+            } else if memflags.table() {
+                self.table
+            } else if memflags.vmctx() {
+                self.vmctx
+            } else {
+                self.other
+            }
+        } else if func.dfg[inst].opcode().can_load() || func.dfg[inst].opcode().can_store() {
+            inst.into()
+        } else {
+            PackedOption::default()
+        }
+    }
+
+    fn meet_from(&mut self, other: &LastStores, loc: Inst) {
+        let meet = |a: PackedOption<Inst>, b: PackedOption<Inst>| -> PackedOption<Inst> {
+            match (a.into(), b.into()) {
+                (None, None) => None.into(),
+                (Some(a), None) => a,
+                (None, Some(b)) => b,
+                (Some(a), Some(b)) if a == b => a,
+                _ => loc.into(),
+            }
+        };
+
+        self.heap = meet(self.heap, other.heap);
+        self.table = meet(self.table, other.table);
+        self.vmctx = meet(self.vmctx, other.vmctx);
+        self.other = meet(self.other, other.other);
+    }
+}
+
+/// A key identifying a unique memory location.
+///
+/// For the result of a load to be equivalent to the result of another
+/// load, or the store data from a store, we need for (i) the
+/// "version" of memory (here ensured by having the same last store
+/// instruction to touch the disjoint category of abstract state we're
+/// accessing); (ii) the address must be the same (here ensured by
+/// having the same SSA value, which doesn't change after computed);
+/// (iii) the offset must be the same; and (iv) the accessed type and
+/// extension mode (e.g., 8-to-32, signed) must be the same.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+struct MemoryLoc {
+    last_store: PackedOption<Inst>,
+    address: Value,
+    offset: Offset32,
+    ty: Type,
```

Comment:
Can we add a test for loads of the same address but of different types? I.e. something like

```
v1 = load.i32 v0
v2 = load.f32 v0
;; check: ... the second load was not eliminated ...
```

Probably would be good to test signed vs unsigned extensions and 8-to-32 vs 16-to-32 extensions as well.

Reply:
Done! Added `filetests/alias/extends.clif`.